### PR TITLE
Upload: accept an InputStream besides a File

### PR DESCRIPTION
**Pull request by alb-i986**
This makes it possible to stream data coming from another server straight into the POST request to Gitlab, saving an extra, expensive, and useless IO operation on disk.

Use case: integrating Gitlab with another defect management system.

### CODE REVIEW

File: src/main/java/org/gitlab4j/api/GitLabApiClient.java [hunk 1]
```diff
@@ -600,23 +600,37 @@ protected Response upload(String name, File fileToUpload, String mediaTypeString
      * @throws IOException if an error occurs while constructing the URL
      */
     protected Response upload(String name, File fileToUpload, String mediaTypeString, Form formData, URL url) throws IOException {
+        MediaType mediaType = (mediaTypeString != null ? MediaType.valueOf(mediaTypeString) : null);
+        FileDataBodyPart filePart = mediaType != null ?
```
Comment (jabby) [inline src/main/java/org/gitlab4j/api/GitLabApiClient.java:604]: After checking in `FileDataBodyPart` constructors, you don't need to test if mediaType is null or not. In fact the constructor used line 606 call the one used line 605 with null in MediaType. Please remove the ternary operator.

File: src/main/java/org/gitlab4j/api/GitLabApiClient.java [hunk 2]
```diff
@@ -600,23 +600,37 @@ protected Response upload(String name, File fileToUpload, String mediaTypeString
...
 
+    protected Response upload(String name, InputStream inputStream, String filename, String mediaTypeString, Object... pathArgs) throws IOException {
+        URL url = getApiUrl(pathArgs);
+        return (upload(name, inputStream, filename, mediaTypeString, null, url));
+    }
+
+    protected Response upload(String name, InputStream inputStream, String filename, String mediaTypeString, Form formData, URL url) throws IOException {
         MediaType mediaType = (mediaTypeString != null ? MediaType.valueOf(mediaTypeString) : null);
-        try (FormDataMultiPart multiPart = new FormDataMultiPart()) {
+        StreamDataBodyPart streamDataBodyPart = mediaType != null ?
```
Comment (jabby) [inline src/main/java/org/gitlab4j/api/GitLabApiClient.java:617]: After checking in `StreamDataBodyPart` constructors, you don't need to test if mediaType is null or not. In fact the constructor used line 619 call the one used line 618 with null in MediaType. Please remove the ternary operator.

**Review by jabby — CHANGES_REQUESTED** [review]
Thank you for your PR and sorry for taking so much time before reviewing your work.

File: src/main/java/org/gitlab4j/api/GitLabApiClient.java (hunk 1 above)
Comment (alb-i986) [reply src/main/java/org/gitlab4j/api/GitLabApiClient.java:604]: Oh yes, I see.
Nice, we can rely on `FileDataBodyPart`'s capability of detecting the media type from a `File`.

The method parameter `mediaTypeString` should be removed then.
Since it's declared as `protected` it is exposed to clients, so we may not remove it until we get to a major version, isn't it?
I'll just update the javadocs for now.

File: src/main/java/org/gitlab4j/api/GitLabApiClient.java (hunk 2 above)
Comment (alb-i986) [reply src/main/java/org/gitlab4j/api/GitLabApiClient.java:617]: In this case it's not really the same thing.
As `StreamDataBodyPart`'s javadoc states:

> It sets the InputStream as a body part with the default MediaType.APPLICATION_OCTET_STREAM_TYPE (if not specified by the user).
> Note: The MIME type of the entity cannot be automatically predicted as in case of FileDataBodyPart.

[Gitlab docs](https://docs.gitlab.com/ee/api/projects.html#upload-a-file) do not help clarifying.
But I guess it's not a problem, I think Gitlab REST API doesn't care about the media type.
What do you think @jabby?

**Review by jabby — APPROVED** [review]

**Comment by alb-i986** [comment]
Thanks a lot!

Cheers
